### PR TITLE
Add new CustomModelDownloadService to retrieve download details about…

### PR DESCRIPTION
**Pull request by annzimmer**
… the CustomModel.

### CODE REVIEW

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+  private static final Charset UTF_8 = StandardCharsets.UTF_8;
+
+  private static final String ETAG_HEADER = "ETag";
+
+  @VisibleForTesting
+  static final String INSTALLATIONS_AUTH_TOKEN_HEADER = "X-Goog-Firebase-Installations-Auth";
+
+  @VisibleForTesting
+  static final String DOWNLOAD_MODEL_REGEX =
+      "https://firebaseml.googleapis.com/Model/v1beta2/projects/%s/models/%s:download";
```
Comment (manjanac) [inline firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java:68]: I think the URL needs to be: "https://firebaseml.googleapis.com/v1beta2/projects/%s/models/%s:download".

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java [hunk 2]
```diff
@@ -42,7 +44,46 @@
...
+   * temporary step towards setting the actual download id.
+   *
+   * @param name - model name
+   * @param modelHash - model hash size
+   * @param fileSize - model file size
+   * @param downloadUrl - download url path
+   * @param downloadUrlExpiry - time download url path expires
+   * @hide
+   */
+  public CustomModel(
```
Comment (rlazo) [inline firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java:80]: Although probably not a big deal, but the order of common args in this constructor and in the prev one doesn't match.

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java [hunk 3]
```diff
@@ -42,7 +44,46 @@
...
+    this(name, downloadId, fileSize, modelHash, "", "", 0);
+  }
+
+  /**
+   * Use when creating a custom model from a stored model with a new download is the background.
+   *
+   * @param name - model name
+   * @param downloadId - Android Download Manger - download id
+   * @param fileSize - model file size
+   * @param modelHash - model hash size
```
Comment (rlazo) [inline firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java:56]: is it the size or the hash itself?

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java [hunk 4]
```diff
@@ -42,7 +44,46 @@
...
+    this(name, downloadId, fileSize, modelHash, localFilePath, "", 0);
+  }
+
+  /**
+   * Use when creating a custom model from a download service response. Download url and download
+   * url expiry should go together. These will not be stored in user preferences as this is a
+   * temporary step towards setting the actual download id.
+   *
+   * @param name - model name
+   * @param modelHash - model hash size
```
Comment (rlazo) [inline firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java:74]: is it the size or the hash itself?

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+  private FirebaseInstallationsApi firebaseInstallations;
+  private String apiKey;
+  private String downloadHost = FIREBASE_DOWNLOAD_HOST;
+
+  public CustomModelDownloadService(
+      FirebaseOptions firebaseOptions, FirebaseInstallationsApi installationsApi) {
+    firebaseInstallations = installationsApi;
+    apiKey = firebaseOptions.getApiKey();
+    // is this an appropriate executor?
+    executorService = Executors.newCachedThreadPool();
```
Comment (rlazo) [inline firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java:90]: sgtm

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+  @VisibleForTesting
+  static long parseTokenExpirationTimestamp(String expiresIn) {
+    if (expiresIn == null || expiresIn.length() == 0) {
+      return 0;
+    }
+
+    try {
+      String isoDatePattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
+      SimpleDateFormat iso8601Format = new SimpleDateFormat(isoDatePattern, Locale.US);
```
Comment (rlazo) [inline firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java:181]: maybe a class level constant?

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    } catch (ParseException pe) {
+      // log error and maybe throw an error
+      Log.w(TAG, "unable to parse datetime:" + expiresIn, pe);
+      return 0;
+    }
+  }
+
+  private Task<CustomModel> fetchDownloadDetails(String modelName, HttpURLConnection connection)
+      throws Exception {
+
```
Comment (rlazo) [inline firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java:194]: unnecessary extra line

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+      // todo(annz) add more specific error handling. NOT_FOUND, etc.
+      return Tasks.forException(
+          new Exception(
+              String.format(
+                  Locale.getDefault(),
+                  "Failed to connect to Firebase ML download server with HTTP status code: %d"
+                      + " and error message: %s",
+                  connection.getResponseCode(),
+                  errorMessage)));
+    } else if (httpResponseCode == HttpURLConnection.HTTP_NOT_MODIFIED) {
```
Comment (rlazo) [inline firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java:211]: conditional could be simpler as

```
if (httpResponseCode == HttpURLConnection.HTTP_OK) {
  return Tasks.forResult(readCustomModelResponse(modelName, connection));
}
if (httpResponseCode == HttpURLConnection.HTTP_NOT_MODIFIED) {
  return Tasks.forResult(null)
}

// handle error

```

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+    String modelHash = maybeUnGzipHeader(connection.getHeaderField(ETAG_HEADER), encodingKey);
+
+    if (modelHash == null || modelHash.isEmpty()) {
+      // todo(annz) replace this...
+      modelHash = connection.getResponseMessage();
+    }
+
+    // JsonReader.peek will sometimes throw AssertionErrors in Android 8.0 and above. See
+    // https://b.corp.google.com/issues/79920590 for details.
```
Comment (rlazo) [inline firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java:236]: Use b/79920590

(other instances fixed in https://github.com/firebase/firebase-android-sdk/pull/2174)

Comment (annzimmer) [reply firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java]: Done.

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java (hunk 8 above)
Comment (annzimmer) [reply firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java]: updated.

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java (hunk 6 above)
Comment (annzimmer) [reply firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java]: done

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java (hunk 2 above)
Comment (annzimmer) [reply firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java:80]: Good catch - made more consistent.

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java [hunk 10]
```diff
@@ -130,19 +177,43 @@ public boolean equals(Object o) {
...
+   *
+   * @hide
+   */
+  public long getDownloadUrlExpiry() {
+    return downloadUrlExpiry;
+  }
+
+  /**
+   * @return the model download url
+   *     <p>Internal use only
```
[the remark this answers is not in the document]
Comment (annzimmer) [reply firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java:204]: formatting adds this back

**Review by rlazo — APPROVED** [review]

**Pull request merged**